packages/fxa-auth-server/lib/oauthdb/index.js

```diff
@@ -99,7 +101,19 @@ module.exports = (log, config) => {
 
     async grantTokensFromAuthorizationCode(oauthParams) {
       try {
-        return await api.grantTokensFromAuthorizationCode(oauthParams);
+        const codeResponse = await api.grantTokensFromAuthorizationCode(
+          oauthParams
+        );
+        const scopeSet = ScopeSet.fromString(codeResponse.scope);
+        if (scopeSet.contains(SESSION_TOKEN_OAUTH_SCOPE)) {
+          // at this point the auth code request has
+          // validated that the session scope is allowed for this client
+          codeResponse.session_token = await db.createSessionToken({
+            uid: codeResponse.uid,
+          });
```

Comment:
This call to `db.createSessionToken` is, I think, the hardest part. The object you pass in here can have [a tremendous number of optional properties](https://github.com/mozilla/fxa/blob/master/packages/fxa-auth-server/lib/tokens/session_token.js#L23) which affect the security of the system.

In general we need to make sure that the sessionToken created here is not any more powerful than the one that was used to create the authorization code. The [`/session/duplicate` route](https://github.com/mozilla/fxa/blob/dfc7a258402840e35139fd918c7bc93bc39ca9b7/packages/fxa-auth-server/lib/routes/session.js#L272) may be a good reference point for the sorts of things we need to consider here from a security perspective. In fact, "return a duplicate of the sessionToken that granted to code" is probably a good way to frame the requirements here.

Some thoughts:

* You're creating a *verified* sessionToken here. Is that appropriate? I think yes, but we'll need to ensure that only verified tokens can be used to create codes with `tokens/session` scope, and I don't think that's true right now.
* If we do create verified tokens, to what should we set the `verificationMethod` and `verifiedAt` properties? Should they be copied from the originating sessionToken?
* What should the `createdAt` and `authAt` be on this token? It's important that the last-auth-at time of the new token does not exceed the last-auth-at time of the token from which it was generated, otherwise an attacker could use this to get around security measures based checking whether the user recently entered their password.

Ultimately, I think you will need to somehow know the originating sessionToken in order to correctly create a new one.

(As a side note, the fact that `db.createSessionToken` lets you call the API without explicit answers to these questions seems like a bit of a API footgun, but that's a separate issue...)